modules/post/multi/manage/upload_exec.rb

```diff
@@ -6,55 +6,67 @@
 class MetasploitModule < Msf::Post
   include Msf::Post::File
 
-  def initialize(info={})
-    super( update_info( info,
-      'Name'          => 'Upload and Execute',
-      'Description'   => %q{ Push a file and execute it },
-      'License'       => MSF_LICENSE,
-      'Author'        => [ 'egypt'],
-      'Platform'      => [ 'win','linux','osx' ],
-      'SessionTypes'  => [ 'meterpreter','shell' ]
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'         => 'Upload and Execute',
+      'Description'  => %q{Push a file and execute it.},
+      'Author'       => 'egypt',
+      'License'      => MSF_LICENSE,
+      'Platform'     => ['win', 'unix', 'linux', 'osx', 'bsd', 'solaris'],
+      'SessionTypes' => ['meterpreter', 'shell']
     ))
 
-    register_options(
-      [
-        OptPath.new('LPATH', [true,'Local file path to upload and execute']),
-        OptString.new('RPATH', [false,'Remote file path on target (default is basename of LPATH)']),
-      ])
+    register_options([
+      OptPath.new('LPATH',   [true, 'Local file path to upload and execute']),
+      OptString.new('RPATH', [false, 'Remote file path on target (default is basename of LPATH)']),
+      OptString.new('ARGS',  [false, 'Command-line arguments to pass to the uploaded file']),
+      OptInt.new('TIMEOUT',  [true, 'Timeout for command execution', 15])
+    ])
   end
 
-  def rpath
-    if datastore['RPATH'].blank?
-      remote_name = File.basename(datastore['LPATH'])
+  def run
+    upload_file(rpath, lpath)
+
+    if session.platform == 'windows'
+      # Don't use cmd.exe /c start so we can fetch output
+      cmd = rpath
+    else
+      begin
+        # client is an alias for session
+        client.fs.file.chmod(rpath, 0700)
```

Comment:
Maybe we should add a helper function for this, e.g:
https://github.com/wvu-r7/metasploit-framework/pull/18

Reply:
I would prefer if this method could take an octal mode instead of being limited to `700`. Maybe `700` is a fine default until we support additive symbolic modes.